File: algorithm/go/quicksort/quick_sort.go
package quicksort

import (
	"time"
	"math/rand"
)

func QuickSort(arr []int) {
	quickSort(arr, 0, len(arr) - 1)
}

func quickSort(arr []int, l, r int) {
	if l >= r {
		return
	}

	mid := partition(arr, l, r)
	quickSort(arr, l, mid-1)
	quickSort(arr, mid+1, r)
}

func partition(arr []int, l, r int) int {
	rand.Seed(time.Now().Unix())
	swap(arr, l, rand.Int() % (r-l+1) + l)

	pivot := arr[l]
	lo := l + 1
	hi := r

	for {
		for ; lo <= r && arr[lo] < pivot; lo++ {}
		for ; hi >= l+1 && arr[hi] > pivot; hi-- {}
		if lo > hi {
			break
		}
		swap(arr, lo, hi)
		lo++
		hi--
	}

	swap(arr, l, lo-1)
	return lo-1
}

func swap(arr []int, i, j int) {
	temp := arr[j]
	arr[j] = arr[i]
	arr[i] = temp
}

File: algorithm/js/sort/quickSort/quickSort.js
function quickSort(arr) {
  $quickSort(arr, 0, arr.length-1);
}

function $quickSort(arr, l, r) {
  if (l >= r) {
    return;
  }

  const mid = partition(arr, l, r);
  $quickSort(arr, l, mid-1);
  $quickSort(arr, mid+1, r);
}

function partition(arr, l, r) {
  swap(arr, l, parseInt(Math.random() * 10000 % (r - l + 1)) + l);
  const pivot = arr[l];

  let lo = l + 1,
    hi = r;
  
  while(lo <= hi) {
    if (arr[lo] < pivot) {
      lo++;
    } else if (arr[hi] > pivot) {
      hi--;
    } else {
      swap(arr, lo, hi);
      lo++;
      hi--;
    }
  }

  swap(arr, l, lo - 1);
  return lo - 1;
}

function swap(arr, i, j) {
  const temp = arr[j];
  arr[j] = arr[i];
  arr[i] = temp;
}

const arr = Array(10);

for (let i = 0; i < 10; i++) {
  arr[i] = parseInt(Math.random() * 10000);
}

console.time("Quick Sort");
quickSort(arr);
console.timeEnd("Quick Sort");

for (let i = 0; i < 9; i++) {
  if (arr[i] > arr[i+1]) {
    throw new Error("Error");
  }
}